<a href="https://colab.research.google.com/github/jokercsi/Investment-Portfolio-Optimization/blob/main/BertTokenizer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

BERT 를 사용하기 위한 패키지

In [1]:
pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 5.3 MB 5.2 MB/s 
     |████████████████████████████████| 7.6 MB 50.0 MB/s 
     |████████████████████████████████| 163 kB 38.4 MB/s 


transformers 패키지를 사용하여 BERT 서브워드 토크나이저를 사용하기

In [2]:
import pandas as pd
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased") # Bert-base의 토크나이저

Downloading:   0%|          | 0.00/232k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/570 [00:00<?, ?B/s]

'Here is the sentence I want embeddings for.'라는 문장을 BERT의 토크나이저로 토큰화 하기

In [3]:
result = tokenizer.tokenize('Here is the sentence I want embeddings for.')
print(result)

['here', 'is', 'the', 'sentence', 'i', 'want', 'em', '##bed', '##ding', '##s', 'for', '.']


embeddings라는 단어는 단어 집합에 존재하지 않으므로 em, ##bed, ##ding, #s로 분리됩니다. 실제로 BERT의 단어 집합에 특정 단어가 있는지 조회하려면 .vocab[]을 통해서 가능합니다. 우선 단어 'here'을 조회해봅시다.

In [4]:
print(tokenizer.vocab['here'])

2182


정수 2182가 출력됩니다. 이는 단어 here이 정수 인코딩을 위해서 단어 집합 내부적으로 2182라는 정수로 맵핑되어져 있다는 의미입니다. 단어 embeddings를 조회해봅시다.

In [5]:
print(tokenizer.vocab['embeddings'])

KeyError: ignored

해당 단어가 존재하지 않는다는 의미에서 KeyError가 발생합니다. 반면, 단어 em, ##bed, ##ing, ##s는 모두 단어 집합에 존재합니다.

In [ ]:
print(tokenizer.vocab['em'])

In [ ]:
print(tokenizer.vocab['##bed'])

In [ ]:
print(tokenizer.vocab['##ding'])

In [ ]:
print(tokenizer.vocab['##s'])

이번에는 BERT의 단어 집합 전체를 불러와서 살펴보겠습니다. BERT의 단어 집합을 vocabulary.txt에 저장합니다.

In [ ]:
# BERT의 단어 집합을 vocabulary.txt에 저장
with open('vocabulary.txt', 'w') as f:
  for token in tokenizer.vocab.keys():
    f.write(token + '\n')

vocabulary.txt 파일을 직접 열어서 살펴볼 수도 있겠지만, 데이터프레임 형태로 저장해서 확인할 수도 있습니다.

In [ ]:
df = pd.read_fwf('vocabulary.txt', header=None)
df

In [ ]:
print('단어 집합의 크기 :',len(df))

BERT의 단어 집합의 크기는 30,522입니다. 데이터프레임의 인덱스가 해당 단어와 맵핑된 정수이므로 정수로부터 맵핑된 단어를 확인할 수 있습니다. 가령, 4667번 단어는 ##ding입니다.

In [ ]:
df.loc[4667].values[0]

참고로 BERT에서 사용되는 특별 토큰들과 그와 맵핑되는 정수는 다음과 같습니다.

[PAD] - 0
[UNK] - 100
[CLS] - 101
[SEP] - 102
[MASK] - 103
임의로 102번 토큰만 출력해봅시다.

In [ ]:
df.loc[102].values[0]